Implements: Blogging prompt card viewmodel slice

AjeshRPai opened the pull request:
## Description
This PR implements a ViewModel Slice for the Blogging prompt card. This PR also removes the blogging prompt feature flags. 

## To test:
- Go to My Site dashboard 
- Test all the features in Blogging prompt card 
- Verify that it works as expected 
- Test all the features related to Blogging prompt

## Regression Notes
1. Potential unintended areas of impact
Blogging prompt card and Blogging prompt feature is working as expected

2. What I did to test those areas of impact (or what existing automated tests did I relied on)
Manual testing 

3. What automated tests I added (or what prevented me from doing so)
Unit tests and Manual testing

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] Talkback.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] Large and small screen sizes. (Tablet and smaller phones)
- [x] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

AjeshRPai committed:
[WIP] + Adds: Viewmodel slice for blogging prompt card

AjeshRPai committed:
[WIP] + Adds: Blogging prompt card navigation to SiteNavigationAction

AjeshRPai committed:
[WIP] + Adds: Blogging prompt card - handle navigation action

AjeshRPai committed:
- Removes: live data for each blogging prompt card action

AjeshRPai committed:
↑ Updates: the naming of Blogging Navigation action classess

AjeshRPai committed:
_ Removes: the dependancy of Blogging feature flags from implementation

[WIP]: the feature flags hasnt been removed, will be removed in future
PR
- Removes: the member variable in builder params and card data class
to toggle on and off the feature

AjeshRPai committed:
→ Moves: the logic of building the Blogging Prompt card

→ Moves: the logic of building the Blogging Prompt card
to  BloggingPromptCardViewModelSlice

AjeshRPai committed:
→ Moves: BloggingPromptsCardTrackHelper logic to ViewModelSlice

AjeshRPai committed:
↑ Updates: the blogging prompt undo click

AjeshRPai committed:
↑ Updates: color res annotation

AjeshRPai committed:
_ Removes: unused import

AjeshRPai committed:
_ Removes: redundant tests

AjeshRPai committed:
* Fixes: failing Cards Builder tests

AjeshRPai committed:
↑ Updates: the blogging prompt undo click private

AjeshRPai committed:
+ Adds: BloggingPromptCardViewModelSliceTest

AjeshRPai committed:
* Fixes: MySiteViewModelTest

AjeshRPai committed:
Merge branch 'issue/move-site-items-builder-to-viewmodel-slice' into issue/move-blogging-prompt-card-items-to-viewmodel-slice

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt
#	WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt

AjeshRPai committed:
* Fixes: Unused imports

AjeshRPai committed:
* Fixes: label color type value

AjeshRPai committed:
- Removes: resource attribute annotation

AjeshRPai committed:
- Removes: redundant import

AjeshRPai committed:
- Removes: BloggingPromptsSocialFeatureConfig flag and updates the usage

AjeshRPai committed:
- Removes: BloggingPromptsListFeatureConfig.kt

AjeshRPai committed:
- Removes: BloggingPromptsEnhancementsFeatureConfig.kt

- Removes: the feature flag
↑ Updates: the feature flag to a helper class to determine when the
feature should be enabled/disabled

AjeshRPai committed:
- Removes: BLOGGING_PROMPTS feature flag

- Removes: the feature flag
↑ Updates: the feature flag to a helper class to determine when the
feature should be enabled/disabled

AjeshRPai committed:
- Renames: BloggingPromptsFeatureConfig.kt to BloggingPromptsFeature

AjeshRPai committed:
* Fixes: Lint errors

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/bloggingprompts/BloggingPromptCardViewModelSlice.kt:
```diff
@@ -1,4 +1,114 @@
+    val onSnackbarMessage = _onSnackbarMessage as LiveData<Event<SnackbarMessageHolder>>
+
+    private val _onNavigation = MutableLiveData<Event<SiteNavigationAction>>()
+    val onNavigation = _onNavigation
```
💡 This seems to be missing a `as LiveData<Event<SiteNavigationAction>>()` if we want to avoid exposing the internal `MutableLiveData` type.

AjeshRPai replied:
Nice catch. Agree 👍🏼 . Done in [cf40c81](https://github.com/wordpress-mobile/WordPress-Android/pull/19152/commits/cf40c8186151e4815807822cc4508b85b56b5cc1)

thomashorta commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/cards/dashboard/bloggingprompts/BloggingPromptCardViewModelSliceTest.kt:
```diff
@@ -0,0 +1,225 @@
+    private val siteLocalId = 1
+    private val siteUrl = "http://site.com"
+    private val siteIcon = "http://site.com/icon.jpg"
+    private val siteName = "Site"
```
💡 It seems these are only being used once to initialize a fake `SiteModel` (`site` field) inside the `setup` phase of the test. I believe this could be hard coded to the class initialization by making something like this:

```suggestion
    private val site: SiteModel = SiteModel().apply {
        id = 1
        siteId = 1L
        url = "http://site.com"
        name = "Site"
        iconUrl = "http://site.com/icon.jpg"
    }
```

And then removing lines 77-81 inside the `setup`.

The main benefit is that this setup, which can safely be shared between test cases, will not run for each test, and it also makes the `setup` function cleaner.

AjeshRPai replied:
I agree 👍🏼 . Thanks for this suggestion. I have made changes to the site model initialization with this commit 42a8d1e0d99f1b827fd48be45a9cae9351be93f5

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsSettingsHelper.kt:
```diff
@@ -48,21 +48,21 @@ class BloggingPromptsSettingsHelper @Inject constructor(
 
         // if the enhancements is turned off, consider the prompts user-enabled, otherwise check the user setting
-        val isPromptsSettingUserEnabled = !bloggingPromptsEnhancementsFeatureConfig.isEnabled() ||
+        val isPromptsSettingUserEnabled = !bloggingPromptsEnhancementsFeature.isEnabled() ||
```
⚠️ Now that this `bloggingPromptsEnhancementsFeature` has the exact same condition to be enabled as `bloggingPromptsFeature` I think this check doesn't make sense anymore, we should only check the `isPromtSettingEnabled` in the final return logic below.

AjeshRPai replied:
👍🏼 Done in e373f8e785efd76e3418c2ff94a8dfdfcd92a354 ✅

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsSettingsHelper.kt:
```diff
@@ -48,21 +48,21 @@ class BloggingPromptsSettingsHelper @Inject constructor(
 
     fun shouldShowPromptsSetting(): Boolean {
-        return isPromptsFeatureAvailable() && bloggingPromptsEnhancementsFeatureConfig.isEnabled()
+        return isPromptsFeatureAvailable() && bloggingPromptsEnhancementsFeature.isEnabled()
```
⚠️ Similar issue here. I am pretty sure we can get rid of the `bloggingPromptsEnhancementsFeature` check since `isPromptsFeatureAvailable` already checks the `bloggingPromptsFeaure`, which has the same condition to be enabled (i.e.: Jetpack app).

AjeshRPai replied:
👍🏼 Done in e373f8e785efd76e3418c2ff94a8dfdfcd92a354 ✅

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/bloggingprompts/BloggingPromptCardViewModelSlice.kt:
❤️ Really nice to have a `Slice` of the ViewModel responsible for a particular part of the screen, especially for the dashboard cards. Really great refactor, thank you!

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/bloggingprompts/BloggingPromptCardViewModelSlice.kt:
```diff
@@ -0,0 +1,132 @@
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
```
⚠️ I don't like this being a `Singleton` especially because it could hold some state. Right now it only holds state in the form of "Event" `LiveData`, which can (in theory) only be consumed once, so this should currently be fine and working as expected. 

The problem, for me, is that it gives a false impression that I could add other regular `LiveData` and/or `Flow` here, but doing that could cause leaking state/data between sites or even accounts (if we don't manually clear stuff).

❓ Is the `Singleton` here needed for something in particular? Maybe I'm missing something.

AjeshRPai replied:
Hey @thomashorta 

Although singleton is not gonna make any issues in this instance. I agree with you, its better to remove it now 👍🏼 .  

Done in [41aeddb](https://github.com/wordpress-mobile/WordPress-Android/pull/19152/commits/41aeddb0649554b579029e96d5cfb5e8a7b713ce)

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/bloggingprompts/BloggingPromptCardViewModelSlice.kt:
```diff
@@ -0,0 +1,132 @@
+    private val bloggingPromptsCardAnalyticsTracker: BloggingPromptsCardAnalyticsTracker,
+    private val bloggingPromptsSettingsHelper: BloggingPromptsSettingsHelper,
+    private val bloggingPromptsCardTrackHelper: BloggingPromptsCardTrackHelper
+) : ScopedViewModel(mainDispatcher) {
```
🛑 Blocker: The `Slice` should NOT be a real Android `ViewModel` as this has some implications that are not properly fulfilled when we use it as a regular POJO (as opposed to properly initializing it via `ViewModelProvider` or other injection functions that use it).

The main one that I am concerned about is the fact that `ViewModels` are "clearable" and they are automatically cleared by the Lifecycle framework when properly attached to lifecycle contexts (e.g.: a `ViewModel` provided to an `Activity` will be cleared and removed from memory if it's parent `Activity` dies). This will NOT happen to this `ViewModel` `Slice`, since it 1. is a Singleton in the dependency graph, and 2. is not attached to a lifecycle.

One of the things that are disposed of/closed is the `viewModelScope`, which would NOT be canceled for this `Slice`, making the coroutines started by it continue to run even if the screen that uses it dies.

My suggestion would be not to make this a subclass of `ScopedViewModel` and instead provide a `CoroutineScope`to it by calling some kind of initializer on it. I saw a somewhat good approach for it, in this context of `ViewModel` slices in this post by Rodrigo Lino: https://medium.com/@rodrigolino/slicing-your-viewmodel-with-delegates-2594e6ab4959

AjeshRPai replied:
👍🏼  Yeah Makes sense 👍🏼 . 

>One of the things that are disposed of/closed is the viewModelScope, which would NOT be canceled for this Slice, making the coroutines started by it continue to run even if the screen that uses it dies.

Yeah 👍🏼  Correct

>My suggestion would be not to make this a subclass of ScopedViewModel and instead provide a CoroutineScopeto it by calling some kind of initializer on it. I saw a somewhat good approach for it, in this context of ViewModel slices in this post by Rodrigo Lino: https://medium.com/@rodrigolino/slicing-your-viewmodel-with-delegates-2594e6ab4959

Thanks for the link @thomashorta. I have updated the Viewmodel Slice to have an initialization function to get the scope. BloggingPromptCardViewModelSlice is no longer a subclass of `ScopedViewModel`. 

Done in [41aeddb](https://github.com/wordpress-mobile/WordPress-Android/pull/19152/commits/41aeddb0649554b579029e96d5cfb5e8a7b713ce)

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/promptslist/BloggingPromptsListViewModel.kt:
```diff
@@ -57,7 +57,7 @@ class BloggingPromptsListViewModel @Inject constructor(
     fun onPromptListItemClicked(itemModel: BloggingPromptsListItemModel) {
         tracker.trackItemClicked()
-        if (bloggingPromptsEnhancementsFeatureConfig.isEnabled()) {
+        if (bloggingPromptsEnhancementsFeature.isEnabled()) {
```
💡 You can probably get rid of this check, since the item that the user clicks will only be visible if the Blogging Prompts Feature is enabled (i.e.: in Jetpack app), which is the exact same condition as this check.

AjeshRPai replied:
👍🏼 Done in e373f8e785efd76e3418c2ff94a8dfdfcd92a354 ✅

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/onboarding/BloggingPromptsOnboardingUiStateMapper.kt:
💡 Similar to what was done in `BloggingPromptCardBuilder.kt`, this class likely doesn't need to check the `BloggingPromptsSocialFeature` anymore, since the mapping only happens when `BloggingPrompts` feature is enabled (i.e.: Jetpack app) and that feature has the exact same condition as the `Social` feature, so it is guaranteed that this will also be `true` if the methods from this class are called.

AjeshRPai replied:
👍🏼 Done in https://github.com/wordpress-mobile/WordPress-Android/commit/3d6ee129b6a2875321a4bcc832beb9f2670b5d94 ✅

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:
❤️ Thanks for reducing the footprint of this class, it really needed it! 🙇🏼

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/SiteNavigationAction.kt:
```diff
@@ -104,3 +105,13 @@ sealed class SiteNavigationAction {
     data class OpenDomainTransferPage(val url: String) : SiteNavigationAction()
 }
+
+sealed class BloggingPromptCardNavigationAction: SiteNavigationAction() {
```
❤️ That makes a lot of sense, way better to have these centralized Navigation Actions!

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/tabs/MySiteTabFragment.kt:
```diff
@@ -479,6 +435,57 @@ class MySiteTabFragment : Fragment(R.layout.my_site_tab_fragment),
+        is BloggingPromptCardNavigationAction -> handleNavigation(action)
+    }
+
+    private fun handleNavigation(action: BloggingPromptCardNavigationAction) {
```
🔍 Nitpick: The name of this method is pretty generic. I would personally prefer something like `handleBloggingPromptNavigation`, but I understand that the param type is also part of the method signature so I am fine if you change it or keep it like that.

AjeshRPai replied:
Hey @thomashorta I had named the function initially `handleBloggingPromptNavigation` but considering the param type I renamed it to `handleNavigation. I would like to keep it that way, if thats ok

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditorBloggingPromptsViewModel.kt:
```diff
@@ -18,7 +18,7 @@ class EditorBloggingPromptsViewModel
     private val bloggingPromptsStore: BloggingPromptsStore,
     private val bloggingPromptsEditorBlockMapper: BloggingPromptsEditorBlockMapper,
-    private val bloggingPromptsEnhancementsFeatureConfig: BloggingPromptsEnhancementsFeatureConfig,
+    private val bloggingPromptsEnhancementsFeature: BloggingPromptsEnhancementsFeature,
```
💡 I believe this is not needed anymore as it will always be true for the Jetpack app and this VM is only started in case the Prompt feature is on and the editor is being started from the Prompts card, so right now the `BloggingPromptsEnhancementsFeature` is **always** `true` when it's used in this class.

AjeshRPai replied:
👍🏼 Done in e373f8e785efd76e3418c2ff94a8dfdfcd92a354

thomashorta commented on WordPress/src/main/java/org/wordpress/android/util/config/BloggingPromptsEnhancementsFeature.kt:
```diff
@@ -0,0 +1,11 @@
+import org.wordpress.android.BuildConfig
+import javax.inject.Inject
+
+class BloggingPromptsEnhancementsFeature
```
💡 From my comments above, I believe this `Feature` is always enabled in the contexts it's used, since we always have a previous check somewhere for the `BloggingPromptsFeature`, which has the exact same logic as this. I suggest removing this class altogether.

AjeshRPai replied:
👍🏼  Thanks @thomashorta . I have removed the class altogether in commit e373f8e785efd76e3418c2ff94a8dfdfcd92a354.

thomashorta commented on WordPress/src/main/java/org/wordpress/android/util/config/BloggingPromptsSocialFeature.kt:
```diff
@@ -0,0 +1,10 @@
+import org.wordpress.android.BuildConfig
+import javax.inject.Inject
+
+class BloggingPromptsSocialFeature @Inject constructor()  {
```
💡 From my comments above, I believe this `Feature` is always enabled in the contexts it's used, since we always have a previous check somewhere for the `BloggingPromptsFeature`, which has the exact same logic as this. I suggest removing this class altogether.

AjeshRPai replied:
👍🏼 Done in 3d6ee12 ✅

thomashorta commented on WordPress/src/main/java/org/wordpress/android/util/config/BloggingPromptsFeature.kt:
```diff
@@ -0,0 +1,10 @@
+import org.wordpress.android.BuildConfig
+import javax.inject.Inject
+
+class BloggingPromptsFeature @Inject constructor() {
```
❓ I am a bit unsure about making this a regular class instead of still being a Feature Flag. I understand we should get rid of feature flags when we are not using them anymore, but I wonder if it would make sense for this to still be a regular `FeatureConfig` that can be remotely enabled/disabled. Wdyt?

My reason is that I personally think there are 2 kinds of FFs:
1. Longstanding Feature Flags, which let us show/hide a whole feature based on a remote config. Those are useful to hide features when they malfunction due to external causes (e.g.: API issues) and also when deprecating/deleting features that are already in production (e.g.: even users on older versions would have the feature removed).
2. Development/Rollout/Rollback Feature Flags, which are used during development so we can merge incomplete work in the trunk and enable it when we are ready to ship/rollout (or do A/B testing), and also act as a safeguard to rollback a feature if we see issues when it's deployed.

I agree the second kind should be temporary, but I think it might be good to have some of the first kind as well. Not 100% sure though.

AjeshRPai replied:
Hey @thomashorta 
I am unsure about this particular feature flag, but I think removing the feature flag once the feature is in production for 2-3 releases is a good practice. We now have around 50 feature flags in the app, which feels too much. I feel that removing the ones that are not in use would make the code and testing much more straightforward. 

>Longstanding Feature Flags let us show/hide a whole feature based on a remote config. Those are useful to hide features when they malfunction due to external causes (e.g.: API issues) 

Longstanding Feature Flags make sense in the scenario of issues arising due to external causes, but I don't think in the case of Blogging Prompt, I don't think there are any, So I would suggest removing this feature flag. 
 

>and also when deprecating/deleting features that are already in production (e.g.: even users on older versions would have the feature removed)

I fear if we consider this case, we will never remove feature flags 😅 . I don't recall an instance in WordPress where we had to turn off a feature after 2-3 releases, so this is a rare case. 

Having said all this, @thomashorta, if you feel 💯 about keeping this feature flag, you can revert the changes I have made for the Blogging prompt feature flag 👍🏼

thomashorta replied:
> I don't recall an instance in WordPress where we had to turn off a feature after 2-3 releases, so this is a rare case.

Yeah, I agree with you. I think it's safe to remove it.

thomashorta reviewed (changes requested):
Really nice change overall. This refactor is definitely needed!
Thank you a lot for the work! ❤️ 🙇🏼 

I added some suggestions and nitpicks and a couple of blocker comments. Feel free to ping me or ask questions if needed!

AjeshRPai committed:
- Removes: BloggingPromptsEnhancementsFeature and its usages

- Removes: BloggingPromptsEnhancementsFeature as it is always true
- Removes: the feature flag off tests

AjeshRPai committed:
- Removes: BloggingPromptsSocialFeature.kt and its usages

- Removes: BloggingPromptsSocialFeature dependencies
as it is always true
- Removes: the feature flag off tests

AjeshRPai committed:
+ Adds: as live data to avoid exposing the internal livedata

AjeshRPai committed:
↑ Updates: fake site model initialization

AjeshRPai committed:
↑ Updates: BloggingPromptCardViewModelSlice

- Removes: the ScopedViewModel base for BloggingPromptCardViewModelSlice
↑ Updates: the BloggingPromptCardViewModelSlice to hold scope of the
viewmodel

AjeshRPai committed:
↑ Updates: MySiteSourceManager to be Singleton

AjeshRPai committed:
* Fixes: BloggingPromptCardViewModelSliceTest

AjeshRPai commented:
Hey @thomashorta 
 This PR is ready for another look.

⚠️ Also, Please note that with this commit [69ab8ed](https://github.com/wordpress-mobile/WordPress-Android/pull/19152/commits/69ab8edc71bb0cb4b003b31b99703fbf15abb9c6), I have made `MySiteSourceManager` to be a singleton. I did so when I saw that the Blogging Prompt card was not refreshed when I turned off the prompts. I tested this change with multiple sites, and everything looks good. I am pointing out this change so that you are aware of this change and can find any issues I might have missed. 

I considered an alternative approach for making the refresh work by exposing refresh as live data and then calling it from `MySiteTabFragment.` Similar to the one we have implemented in `PostsCardViewModelSlice`, this might take some more time. I will return to the Refresh of data and fetching of data of the ViewModel slices in the future; for now, I think making `MySiteSourceManager` a singleton class is fine. Let me know your thoughts.

thomashorta reviewed (changes requested):
> ⚠️ Also, Please note that with this commit https://github.com/wordpress-mobile/WordPress-Android/pull/19152/commits/69ab8edc71bb0cb4b003b31b99703fbf15abb9c6, I have made `MySiteSourceManager` to be a singleton. I did so when I saw that the Blogging Prompt card was not refreshed when I turned off the prompts. I tested this change with multiple sites, and everything looks good. I am pointing out this change so that you are aware of this change and can find any issues I might have missed.

🛑  I tested the new change and it seems the dashboard loading is broken when this class is a Singleton. I get an infinite progress indicator almost every time after doing something like posting now. Even if I go to `Menu` tab and back to `Home` after that infinite loading it doesn't let me pull to refresh anymore (see video below). This is a **blocker** in my opinion.

https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/0a8f2cce-a9d0-46b4-9966-344ad37746c4

I tested removing the `Singleton` again from that class (basically reverting that commit) and this is fixed again, so I don't get the infinite dashboard progress, but then I see the same behavior you mentioned about the card not hiding when turning the card off from its overflow menu. 

I think the issue here is that `MySiteSourceManager` doesn't really work correctly when it's a Singleton but it also shouldn't be injected into multiple classes via Dagger when it's **not** a Singleton.

What do you think about injecting it via the `initialize` method you created for the scope? That way you can make sure it is always initialized with the same instance used by the `ViewModel`. I tested it quickly and it seems to work.

```kotlin
private lateinit var scope: CoroutineScope
private lateinit var mySiteSourceManager: MySiteSourceManager

fun initialize(scope: CoroutineScope, mySiteSourceManager: MySiteSourceManager) {
    this.scope = scope
    this.mySiteSourceManager = mySiteSourceManager
}
```

Commit:
↑ Updates: MySiteSourceManager to not be Singleton

Commit:
↑ Updates: BloggingPromptCardViewModelSlice

↑ Updates: inject MySourceManager instance via initializer method

Commit:
* Fixes: BloggingPromptCardViewModelSliceTest

thomashorta commented:
As we talked earlier today, @AjeshRPai, I am pushing the changes I suggested in my last comment and approving the PR. When the CI is finished running I will also merge it!

thomashorta reviewed (approved):
I pushed the suggestions from my last review, so the `MySiteSourceManager` is not `Singleton` anymore and the infinite loading is fixed while still being able to hide the Prompts card properly.

Thanks for all the changes! 🙇🏼

AjeshRPai committed:
* Fixes: BloggingPromptCardViewModelSliceTest
